datafusion/physical-plan/src/windows/bounded_window_agg_exec.rs

```diff
@@ -1500,129 +1494,6 @@ mod tests {
         Ok(source)
     }
 
-    // Tests NTH_VALUE(negative index) with memoize feature.
-    // To be able to trigger memoize feature for NTH_VALUE we need to
-    // - feed BoundedWindowAggExec with batch stream data.
-    // - Window frame should contain UNBOUNDED PRECEDING.
-    // It hard to ensure these conditions are met, from the sql query.
-    #[tokio::test]
-    async fn test_window_nth_value_bounded_memoize() -> Result<()> {
```

Comment:
This test should not be removed either.

Reply:
same as: https://github.com/apache/datafusion/pull/13201#discussion_r1835785085

I could not build a valid physical layer window expression that would work. Can I get help on this too if possible @jcsherin @berkaysynnada ? An example would be great

Reply:
I reused `create_udwf_window_expr` by changing the visibility to current module to build the physical expressions required for this test. 

You may apply this patch as-is and the test will pass.

```patch
diff --git a/datafusion/physical-plan/src/windows/bounded_window_agg_exec.rs b/datafusion/physical-plan/src/windows/bounded_window_agg_exec.rs
index 0a898bd85..a031253e4 100644
--- a/datafusion/physical-plan/src/windows/bounded_window_agg_exec.rs
+++ b/datafusion/physical-plan/src/windows/bounded_window_agg_exec.rs
@@ -1163,7 +1163,9 @@ mod tests {
     use crate::expressions::PhysicalSortExpr;
     use crate::projection::ProjectionExec;
     use crate::streaming::{PartitionStream, StreamingTableExec};
-    use crate::windows::{create_window_expr, BoundedWindowAggExec, InputOrderMode};
+    use crate::windows::{
+        create_udwf_window_expr, create_window_expr, BoundedWindowAggExec, InputOrderMode,
+    };
     use crate::{execute_stream, get_plan_string, ExecutionPlan};
 
     use arrow_array::builder::{Int64Builder, UInt64Builder};
@@ -1180,13 +1182,14 @@ mod tests {
         WindowFrame, WindowFrameBound, WindowFrameUnits, WindowFunctionDefinition,
     };
     use datafusion_functions_aggregate::count::count_udaf;
-    use datafusion_functions_window::nth_value::first_value_udwf;
     use datafusion_functions_window::nth_value::last_value_udwf;
     use datafusion_functions_window::nth_value::nth_value_udwf;
-    use datafusion_physical_expr::expressions::{col, lit, Column};
+    use datafusion_physical_expr::expressions::{col, Column, Literal};
     use datafusion_physical_expr::{LexOrdering, PhysicalExpr};
 
     use crate::memory::MemoryExec;
+    use crate::common::collect;
+    use datafusion_physical_expr::window::BuiltInWindowExpr;
     use futures::future::Shared;
     use futures::{pin_mut, ready, FutureExt, Stream, StreamExt};
     use itertools::Itertools;
@@ -1524,36 +1527,35 @@ mod tests {
         )
         .map(|e| Arc::new(e) as Arc<dyn ExecutionPlan>)?;
         let col_a = col("a", &schema)?;
-        let nth_value_func1 = NthValue::nth(
-            "nth_value(-1)",
-            Arc::clone(&col_a),
-            DataType::Int32,
-            1,
+        let nth_value_func1 = create_udwf_window_expr(
+            &nth_value_udwf(),
+            &[Arc::clone(&col_a), Arc::new(Literal::new(ScalarValue::Int32(Some(1))))],
+            &schema,
+            "nth_value(-1)".to_string(),
             false,
         )?
         .reverse_expr()
         .unwrap();
-        let nth_value_func2 = NthValue::nth(
-            "nth_value(-2)",
-            Arc::clone(&col_a),
-            DataType::Int32,
-            2,
+        let nth_value_func2 = create_udwf_window_expr(
+            &nth_value_udwf(),
+            &[Arc::clone(&col_a), Arc::new(Literal::new(ScalarValue::Int32(Some(2))))],
+            &schema,
+            "nth_value(-2)".to_string(),
             false,
-        )?
-        .reverse_expr()
-        .unwrap();
-        let last_value_func = Arc::new(NthValue::last(
-            "last",
-            Arc::clone(&col_a),
-            DataType::Int32,
+        )?.reverse_expr().unwrap();
+        let last_value_func = create_udwf_window_expr(
+            &last_value_udwf(),
+            &[Arc::clone(&col_a)],
+            &schema,
+            "last".to_string(),
             false,
-        )) as _;
+        )?;
         let window_exprs = vec![
             // LAST_VALUE(a)
             Arc::new(BuiltInWindowExpr::new(
                 last_value_func,
                 &[],
-                &LexOrdering::default(),
+                  &LexOrdering::default(),
                 Arc::new(WindowFrame::new_bounds(
                     WindowFrameUnits::Rows,
                     WindowFrameBound::Preceding(ScalarValue::UInt64(None)),
diff --git a/datafusion/physical-plan/src/windows/mod.rs b/datafusion/physical-plan/src/windows/mod.rs
index aee0de94e..a4630c4f7 100644
--- a/datafusion/physical-plan/src/windows/mod.rs
+++ b/datafusion/physical-plan/src/windows/mod.rs
@@ -158,7 +158,7 @@ fn window_expr_from_aggregate_expr(
 }
 
 /// Creates a `BuiltInWindowFunctionExpr` suitable for a user defined window function
-fn create_udwf_window_expr(
+pub(self) fn create_udwf_window_expr(
     fun: &Arc<WindowUDF>,
     args: &[Arc<dyn PhysicalExpr>],
     input_schema: &Schema,

```

Reply:
Thanks a lot @jcsherin. I will apply this changes and try to wrap up now

Reply:
This test passes now